docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+
+* A lower-level component will be responsible for the direct interaction with
+  the engines and with the various servers needed to perform a sync. It will
+  also have the ultimate responsibility to not cause harm to the service (for
+  example, it will be likely to enforce some kind of rate limiting or ensuring
+  that service requests for backoff are enforced)
+
+Because all application-services engines are written in Rust, it's tempting to
+suggest that this lower-level component also be written in Rust and everything
+"just works", but there are a couple of complications here:
+
+* For iOS, we hope to integrate with older engines which aren't written in
+  Rust, even if iOS does move to the new Sync Manager.
+
+* For Desktop, we hope to start by reusing the existing "sync manager"
+  implemented by Desktop, and start moving individual engines across.
+
+* There may be some cross-crate issues even for the Rust implemented engines.
+  Or more specifically, we'd like to avoid assuming any particular linkage or
+  packaging of Rust implemented engines.
+
+Even with these complications, we expect the high-level component
+to be written in a platform specific language (ie, Kotlin or Swift) and our
+lower-level component to be implemented in Rust and delivered as part of the
+application-services library - but that's not a free-pass.
+
+# The responsibilities of the Sync Manager.
+
+The primary responsibilities of the "high level" portion of the sync manager are:
+
+* Manage all FxA interaction. The low-level component will have a way to
+  communicate auth related problems, but it is the high-level component
+  which takes concrete FxA action.
+
+* Expose all UI for the user to choose what to sync and coordinate this with
+  the low-level component. Note that because these choices can be made on any
+  connected device, these choices must be communicated in both directions.
+
+* Implement timers or any other mechanism to fully implement the "sync
+  scheduler", including any policy decisions such as only syncing on WiFi,
+  etc.
+
+* Implement a UI so the user can "sync now".
+
+* Collect telemetry from the low-level component, probably augment it, then
+  submit it to the telemetry pipeline.
+
+The primary responsibilities of the "low level" portion of the sync manager are:
+
+* Manage the `meta/global`, `crypto/keys` and `info/collections` resources,
+  and interact with each engine as necessary based on the content of these
+  resources.
+
+* Manage interaction with the token server.
+
+* Enforce constraints necessary to ensure the entire ecosystem is not
+  subject to undue load. For example, this component should ignore attempts to
+  sync continiously, or to sync when the services have requested backoff.
+
+* Manage the "clients" collection - we probably can't ignore this any longer,
+  especially for bookmarks (as desktop will send a wipe command on bookmark
+  restore, and things will "be bad" if we don't see that command).
```

Comment:
The client engine will want the FxA device ID, simply so it can populate `fxaDeviceId`. Desktop uses this for device names etc in various places.

It's probably ideal if we could use the fxa device ID *as* the client ID, which I'm leaning towards - in my mind, the only issue with that is that the FxA IDs are more ephemeral (ie, don't survive some password/account related events) - but is that actually a problem? We'd track this and clean up after ourselves (ie, we'd delete our old client record when we noticed the ID change, etc)